### PR TITLE
feat: Adding a PGN-Spy mode

### PR DESCRIPTION
**Pull request by lakinwecker**
This copies the way that PGN-Spy controls stockfish for its analysis. Long term this should go in the analysis-worker IMO, but I don't have time to do it that way right now.

What do we think?

### CODE REVIEW

**Review by guychessclub — APPROVED** [review]

File: uci/cmd.go [hunk 1]
```diff
@@ -327,6 +327,130 @@ func (CmdGo) ProcessResponse(e *Engine) error {
...
+//	lines, rather than fully implemented here, but we have limited time
+type CmdPgnSpy struct {
+	SearchMoves []*chess.Move
+	SearchDepth int
+	MinTime     time.Duration
+	MaxTime     time.Duration
+}
+
+func (cmd CmdPgnSpy) String() string {
+	a := []string{"go"}
```
Comment (incu6us) [inline uci/cmd.go:342]: a := []string{"go", "infinite"}

File: uci/cmd.go [hunk 2]
```diff
@@ -327,6 +327,130 @@ func (CmdGo) ProcessResponse(e *Engine) error {
...
+
+	a = append(a, "infinite")
```
Comment (incu6us) [inline uci/cmd.go:344]: then, you can remove it

File: uci/cmd.go [hunk 3]
```diff
@@ -327,6 +327,130 @@ func (CmdGo) ProcessResponse(e *Engine) error {
...
+			extractInfo(info, cmd, &engineStopped, &reachedDepth, &beingMated, &engineCouldBeStuck)
+			if (reachedDepth || beingMated) && !engineStopped {
+				engineStopped = true
+				e.stop()
+			}
+		} else if strings.HasPrefix(text, "bestmove") {
+			// TODO: this should probably be its own struct with UnmarshalText function
+			parts := strings.Split(text, " ")
+			if len(parts) <= 1 {
+				return errors.New("best move not found " + text)
```
Comment (incu6us) [inline uci/cmd.go:388]: `return fmt.Errorf("best move not found %s", text)`

File: uci/cmd.go [hunk 4]
```diff
@@ -327,6 +327,130 @@ func (CmdGo) ProcessResponse(e *Engine) error {
...
+			results.BestMove = bestMove
+			if len(parts) >= 4 {
+				ponderMove, err := chess.UCINotation{}.Decode(nil, parts[3])
+				if err != nil {
+					return err
+				}
+				results.Ponder = ponderMove
+			}
+			break
+		} else {
```
Comment (incu6us) [inline uci/cmd.go:403]: just `continue` without `else` block?

**Pull request merged**